### ISSUE TITLE
unify how strong-mode flag is enabled in the FE

### ISSUE THREAD

**Issue by sigmundch** [dart-lang/sdk#31009]
We have two ways of specifying strong mode:
 * CompilerOptions.strongMode
 * kernel.TargetFlags.strongMode

options.strongMode is mostly used by IKG today, while targetflags.strongMode is indirectly copied around and used in different parts of the compiler.

We should unify on a single way of doing things.

**Comment by sigmundch** [comment]
I think using options.strongMode yields a simpler public API, since most users of the public API could omit specifying a Target.

The challenge of doing so is that today we expect a Target and Target is built using TargetFlags, so we would have to do some changes related to kernel.Target in order to provide the strong mode flag to the targets from the FE. Two options for doing so:
- remove TargetFlags, but pass strong-mode as an argument to the APIs that care about it (currently the transformation APIs)
- keep the TargetFlags, but change the public API of the FE to take a TargertFactory and not a TargetInstance.

I'm currently inclined to do the latter. It is actually similar to what we used to have in dartk and how these TargetFlags were initially designed to work (see [_TargetBuilder][1]).

[1]: https://github.com/dart-lang/sdk/blob/master/pkg/kernel/lib/target/targets.dart#L38

**Comment by srawlins** [comment]
I think we're past this?

**Comment by sigmundch** [comment]
I agree that for strong-mode there is not much to do, but I'll let @kmillikin close this or rename the bug.

In particular, the issue exists still for all other flags used to configure the CFE: there is no canonical way to specify the flags yet, and we've had issues recently where our setup and implicit defaults made us accidentally disable sync-async in modular kernel-based vm tests in bazel.
